{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing the performance the updated model trained using 1M segments vs 100K segments by Adam with the older tensorflow model trained with 300K segments.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.datasets import load_svmlight_file\n",
    "from sklearn.externals import joblib\n",
    "import argparse\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn.metrics import confusion_matrix, precision_recall_curve\n",
    "from sklearn.metrics import classification_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = pd.read_csv('~/all_2k_prediction.out.label.prediction_combined',sep=' ',header=None,names=['name','length','probability','prediction','vfam','pfam','diamond','label',\n",
    "                                                                                      'score_euk_non_euk','pred_euk_non_euk','score_non_euk_virus','pred_non_euk_virus',\n",
    "                                                                                      'score_virus_non_virus','pred_virus_non_virus'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>probability</th>\n",
       "      <th>prediction</th>\n",
       "      <th>vfam</th>\n",
       "      <th>pfam</th>\n",
       "      <th>diamond</th>\n",
       "      <th>label</th>\n",
       "      <th>score_euk_non_euk</th>\n",
       "      <th>pred_euk_non_euk</th>\n",
       "      <th>score_non_euk_virus</th>\n",
       "      <th>pred_non_euk_virus</th>\n",
       "      <th>score_virus_non_virus</th>\n",
       "      <th>pred_virus_non_virus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_1_length_91451_cov_163.114_g0_i0</td>\n",
       "      <td>91451</td>\n",
       "      <td>0.000375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996004</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001954</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_2_length_74096_cov_303.895_g1_i0</td>\n",
       "      <td>74096</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.995278</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000529</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_3_length_66420_cov_268.59_g1_i1</td>\n",
       "      <td>66420</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996629</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000761</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_4_length_61338_cov_317.585_g2_i0</td>\n",
       "      <td>61338</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.991967</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_5_length_59036_cov_197.701_g0_i1</td>\n",
       "      <td>59036</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.990354</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    name  length  probability  prediction  \\\n",
       "0  NODE_1_length_91451_cov_163.114_g0_i0   91451     0.000375         0.0   \n",
       "1  NODE_2_length_74096_cov_303.895_g1_i0   74096     0.000088         0.0   \n",
       "2   NODE_3_length_66420_cov_268.59_g1_i1   66420     0.000041         0.0   \n",
       "3  NODE_4_length_61338_cov_317.585_g2_i0   61338     0.000056         0.0   \n",
       "4  NODE_5_length_59036_cov_197.701_g0_i1   59036     0.000040         0.0   \n",
       "\n",
       "   vfam  pfam  diamond  label  score_euk_non_euk  pred_euk_non_euk  \\\n",
       "0   0.0   0.0      0.0    0.0           0.000025               0.0   \n",
       "1   0.0   0.0      0.0    0.0           0.000012               0.0   \n",
       "2   0.0   0.0      0.0    0.0           0.000009               0.0   \n",
       "3   0.0   0.0      0.0    0.0           0.000008               0.0   \n",
       "4   0.0   0.0      0.0    0.0           0.000008               0.0   \n",
       "\n",
       "   score_non_euk_virus  pred_non_euk_virus  score_virus_non_virus  \\\n",
       "0             0.996004                 1.0               0.001954   \n",
       "1             0.995278                 1.0               0.000529   \n",
       "2             0.996629                 1.0               0.000761   \n",
       "3             0.991967                 1.0               0.000477   \n",
       "4             0.990354                 1.0               0.000617   \n",
       "\n",
       "   pred_virus_non_virus  \n",
       "0                   0.0  \n",
       "1                   0.0  \n",
       "2                   0.0  \n",
       "3                   0.0  \n",
       "4                   0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import NCBITaxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ncbi.update_taxonomy_database()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi.get_rank([-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "megan_annot = pd.read_csv('./Micromonas/all_2k_contigs.fa.diamond-ex-all_tax_ID.txt',sep='\\t',header=None,names=['name','tax_id'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>tax_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>205535</th>\n",
       "      <td>NODE_1560_length_4417_cov_606.417_g1101_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205536</th>\n",
       "      <td>NODE_6760_length_2430_cov_7.98316_g5072_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205537</th>\n",
       "      <td>NODE_6861_length_2415_cov_11.7419_g5152_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205538</th>\n",
       "      <td>NODE_427_length_4784_cov_28.9581_g309_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205539</th>\n",
       "      <td>NODE_686_length_4207_cov_38.501_g514_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205540</th>\n",
       "      <td>NODE_8282_length_2222_cov_80.5335_g5687_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205541</th>\n",
       "      <td>NODE_11283_length_2608_cov_11.3604_g3375_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205542</th>\n",
       "      <td>NODE_11930_length_2545_cov_6.42169_g8882_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205543</th>\n",
       "      <td>NODE_3052_length_2128_cov_13.2909_g2214_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205544</th>\n",
       "      <td>NODE_7763_length_2937_cov_37.3105_g5700_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205545</th>\n",
       "      <td>NODE_537_length_7587_cov_91.8853_g232_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205546</th>\n",
       "      <td>NODE_905_length_5770_cov_86.7459_g621_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205547</th>\n",
       "      <td>NODE_907_length_5770_cov_79.6618_g621_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205548</th>\n",
       "      <td>NODE_1193_length_5049_cov_32.3909_g841_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205549</th>\n",
       "      <td>NODE_1424_length_4687_cov_34.253_g841_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205550</th>\n",
       "      <td>NODE_1425_length_4687_cov_34.2418_g841_i2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205551</th>\n",
       "      <td>NODE_1426_length_4687_cov_34.2232_g841_i3</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205552</th>\n",
       "      <td>NODE_10606_length_2041_cov_13.4783_g7959_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205553</th>\n",
       "      <td>NODE_3301_length_4554_cov_17.0625_g2286_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205554</th>\n",
       "      <td>NODE_1792_length_5360_cov_28.1088_g1307_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205555</th>\n",
       "      <td>NODE_2469_length_2590_cov_13.2714_g1834_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205556</th>\n",
       "      <td>NODE_10722_length_2053_cov_20.1552_g8270_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205557</th>\n",
       "      <td>NODE_10871_length_2040_cov_13.3295_g8393_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205558</th>\n",
       "      <td>NODE_10993_length_2030_cov_28.5139_g8483_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205559</th>\n",
       "      <td>NODE_10995_length_2030_cov_20.2582_g8485_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205560</th>\n",
       "      <td>NODE_11091_length_2022_cov_339.209_g8556_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205561</th>\n",
       "      <td>NODE_11271_length_2008_cov_14.0548_g8690_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205562</th>\n",
       "      <td>NODE_11289_length_2006_cov_15.1317_g8702_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205563</th>\n",
       "      <td>NODE_43_length_17528_cov_32.0741_g41_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205564</th>\n",
       "      <td>NODE_783_length_7446_cov_147.605_g488_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205875</th>\n",
       "      <td>NODE_252_length_4134_cov_66.1326_g176_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205876</th>\n",
       "      <td>NODE_443_length_3280_cov_59.1315_g313_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205877</th>\n",
       "      <td>NODE_444_length_3280_cov_59.1029_g313_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205878</th>\n",
       "      <td>NODE_445_length_3280_cov_59.0602_g313_i2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205879</th>\n",
       "      <td>NODE_493_length_3175_cov_38.4122_g349_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205880</th>\n",
       "      <td>NODE_601_length_2975_cov_58.6223_g430_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205881</th>\n",
       "      <td>NODE_885_length_2633_cov_46.557_g430_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205882</th>\n",
       "      <td>NODE_995_length_2535_cov_31.3802_g710_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205883</th>\n",
       "      <td>NODE_1076_length_2484_cov_32.732_g766_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205884</th>\n",
       "      <td>NODE_2208_length_2166_cov_12.6627_g1675_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205885</th>\n",
       "      <td>NODE_2360_length_2124_cov_9.29048_g1808_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205886</th>\n",
       "      <td>NODE_2703_length_2035_cov_9.62323_g2060_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205887</th>\n",
       "      <td>NODE_7270_length_2337_cov_14.9246_g5536_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205888</th>\n",
       "      <td>NODE_7364_length_2325_cov_13.5273_g5536_i1</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205889</th>\n",
       "      <td>NODE_7514_length_2305_cov_9.00622_g5740_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205890</th>\n",
       "      <td>NODE_7609_length_2293_cov_10.9263_g5812_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205891</th>\n",
       "      <td>NODE_7943_length_2251_cov_10.5109_g6064_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205892</th>\n",
       "      <td>NODE_7053_length_2605_cov_12.1471_g5204_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205893</th>\n",
       "      <td>NODE_2934_length_2191_cov_13.5791_g2262_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205894</th>\n",
       "      <td>NODE_3043_length_2162_cov_29.9298_g2344_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205895</th>\n",
       "      <td>NODE_3356_length_2092_cov_18.8625_g2596_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205896</th>\n",
       "      <td>NODE_3439_length_2076_cov_14.1677_g2653_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205897</th>\n",
       "      <td>NODE_3460_length_2073_cov_8.21259_g2669_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205898</th>\n",
       "      <td>NODE_4378_length_3392_cov_107.825_g3177_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205899</th>\n",
       "      <td>NODE_4845_length_3215_cov_21.212_g3532_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205900</th>\n",
       "      <td>NODE_3236_length_3948_cov_22.683_g2346_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205901</th>\n",
       "      <td>NODE_3626_length_3727_cov_26.6925_g2627_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205902</th>\n",
       "      <td>NODE_3698_length_3688_cov_32.7085_g2682_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205903</th>\n",
       "      <td>NODE_4165_length_3470_cov_31.1602_g3025_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205904</th>\n",
       "      <td>NODE_21757_length_2016_cov_7.44875_g15176_i0</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>370 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  tax_id\n",
       "205535    NODE_1560_length_4417_cov_606.417_g1101_i0      -2\n",
       "205536    NODE_6760_length_2430_cov_7.98316_g5072_i0      -2\n",
       "205537    NODE_6861_length_2415_cov_11.7419_g5152_i0      -2\n",
       "205538      NODE_427_length_4784_cov_28.9581_g309_i0      -2\n",
       "205539       NODE_686_length_4207_cov_38.501_g514_i0      -2\n",
       "205540    NODE_8282_length_2222_cov_80.5335_g5687_i0      -2\n",
       "205541   NODE_11283_length_2608_cov_11.3604_g3375_i1      -2\n",
       "205542   NODE_11930_length_2545_cov_6.42169_g8882_i0      -2\n",
       "205543    NODE_3052_length_2128_cov_13.2909_g2214_i0      -2\n",
       "205544    NODE_7763_length_2937_cov_37.3105_g5700_i0      -2\n",
       "205545      NODE_537_length_7587_cov_91.8853_g232_i1      -2\n",
       "205546      NODE_905_length_5770_cov_86.7459_g621_i0      -2\n",
       "205547      NODE_907_length_5770_cov_79.6618_g621_i1      -2\n",
       "205548     NODE_1193_length_5049_cov_32.3909_g841_i0      -2\n",
       "205549      NODE_1424_length_4687_cov_34.253_g841_i1      -2\n",
       "205550     NODE_1425_length_4687_cov_34.2418_g841_i2      -2\n",
       "205551     NODE_1426_length_4687_cov_34.2232_g841_i3      -2\n",
       "205552   NODE_10606_length_2041_cov_13.4783_g7959_i0      -2\n",
       "205553    NODE_3301_length_4554_cov_17.0625_g2286_i0      -2\n",
       "205554    NODE_1792_length_5360_cov_28.1088_g1307_i0      -2\n",
       "205555    NODE_2469_length_2590_cov_13.2714_g1834_i0      -2\n",
       "205556   NODE_10722_length_2053_cov_20.1552_g8270_i0      -2\n",
       "205557   NODE_10871_length_2040_cov_13.3295_g8393_i0      -2\n",
       "205558   NODE_10993_length_2030_cov_28.5139_g8483_i0      -2\n",
       "205559   NODE_10995_length_2030_cov_20.2582_g8485_i0      -2\n",
       "205560   NODE_11091_length_2022_cov_339.209_g8556_i0      -2\n",
       "205561   NODE_11271_length_2008_cov_14.0548_g8690_i0      -2\n",
       "205562   NODE_11289_length_2006_cov_15.1317_g8702_i0      -2\n",
       "205563       NODE_43_length_17528_cov_32.0741_g41_i0      -2\n",
       "205564      NODE_783_length_7446_cov_147.605_g488_i1      -2\n",
       "...                                              ...     ...\n",
       "205875      NODE_252_length_4134_cov_66.1326_g176_i0      -2\n",
       "205876      NODE_443_length_3280_cov_59.1315_g313_i0      -2\n",
       "205877      NODE_444_length_3280_cov_59.1029_g313_i1      -2\n",
       "205878      NODE_445_length_3280_cov_59.0602_g313_i2      -2\n",
       "205879      NODE_493_length_3175_cov_38.4122_g349_i0      -2\n",
       "205880      NODE_601_length_2975_cov_58.6223_g430_i0      -2\n",
       "205881       NODE_885_length_2633_cov_46.557_g430_i1      -2\n",
       "205882      NODE_995_length_2535_cov_31.3802_g710_i0      -2\n",
       "205883      NODE_1076_length_2484_cov_32.732_g766_i0      -2\n",
       "205884    NODE_2208_length_2166_cov_12.6627_g1675_i0      -2\n",
       "205885    NODE_2360_length_2124_cov_9.29048_g1808_i0      -2\n",
       "205886    NODE_2703_length_2035_cov_9.62323_g2060_i0      -2\n",
       "205887    NODE_7270_length_2337_cov_14.9246_g5536_i0      -2\n",
       "205888    NODE_7364_length_2325_cov_13.5273_g5536_i1      -2\n",
       "205889    NODE_7514_length_2305_cov_9.00622_g5740_i0      -2\n",
       "205890    NODE_7609_length_2293_cov_10.9263_g5812_i0      -2\n",
       "205891    NODE_7943_length_2251_cov_10.5109_g6064_i0      -2\n",
       "205892    NODE_7053_length_2605_cov_12.1471_g5204_i0      -2\n",
       "205893    NODE_2934_length_2191_cov_13.5791_g2262_i0      -2\n",
       "205894    NODE_3043_length_2162_cov_29.9298_g2344_i0      -2\n",
       "205895    NODE_3356_length_2092_cov_18.8625_g2596_i0      -2\n",
       "205896    NODE_3439_length_2076_cov_14.1677_g2653_i0      -2\n",
       "205897    NODE_3460_length_2073_cov_8.21259_g2669_i0      -2\n",
       "205898    NODE_4378_length_3392_cov_107.825_g3177_i0      -2\n",
       "205899     NODE_4845_length_3215_cov_21.212_g3532_i0      -2\n",
       "205900     NODE_3236_length_3948_cov_22.683_g2346_i0      -2\n",
       "205901    NODE_3626_length_3727_cov_26.6925_g2627_i0      -2\n",
       "205902    NODE_3698_length_3688_cov_32.7085_g2682_i0      -2\n",
       "205903    NODE_4165_length_3470_cov_31.1602_g3025_i0      -2\n",
       "205904  NODE_21757_length_2016_cov_7.44875_g15176_i0      -2\n",
       "\n",
       "[370 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "megan_annot[megan_annot['tax_id']<0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/homes/q/qpzhang/vica/miniconda/envs/vica/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 171193 was translated into 153265\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/global/homes/q/qpzhang/vica/miniconda/envs/vica/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 206350 was translated into 32003\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/global/homes/q/qpzhang/vica/miniconda/envs/vica/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 383630 was translated into 1660158\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "### Using MEGAN annotation to check model performance \n",
    "\n",
    "def get_top_label(tax_id):\n",
    "    try:\n",
    "        lineage = set(ncbi.get_lineage(tax_id))\n",
    "    except:\n",
    "        return \"N/A\"\n",
    "    for i in lineage:\n",
    "        if i == 2759:\n",
    "            return \"Eukaryota\"\n",
    "        elif i == 10239:\n",
    "            return \"Virus\"\n",
    "        elif i == 2157:\n",
    "            return \"Archaea\"\n",
    "        elif i == 2:\n",
    "            return \"Bacteria\"\n",
    "        elif i == 12884:\n",
    "            return \"Viroids\"\n",
    "        elif i == 28384:\n",
    "            return \"other\"\n",
    "        elif i == 12908:\n",
    "            return \"unclassified\"\n",
    "    else:\n",
    "        return \"N/A\"\n",
    "\n",
    "def get_ranks(tax_id):\n",
    "    ranks = ncbi.get_rank([tax_id])\n",
    "    try:\n",
    "        return ranks[tax_id]\n",
    "    except:\n",
    "        return tax_id\n",
    "    \n",
    "def get_name(tax_id):\n",
    "    ranks = ncbi.get_taxid_translator([tax_id])\n",
    "    try:\n",
    "        return ranks[tax_id]\n",
    "    except:\n",
    "        return tax_id\n",
    "\n",
    "# get MEGAN annotation results\n",
    "\n",
    "megan_annot = pd.read_csv('./Micromonas/all_2k_contigs.fa.diamond-ex-all_tax_ID.txt',sep='\\t',header=None,names=['name','tax_id'])\n",
    "\n",
    "megan_annot['rank'] = megan_annot['tax_id'].apply(get_ranks)\n",
    "megan_annot['tax_name']= megan_annot['tax_id'].apply(get_name)\n",
    "megan_annot['top_label']= megan_annot['tax_id'].apply(get_top_label)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['no rank']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(ncbi.get_rank([1]).values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tax_name</th>\n",
       "      <th>top_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_646_length_5889_cov_28.079_g448_i0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_667_length_5857_cov_28.2491_g448_i1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_671_length_5843_cov_18.2623_g465_i0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_806_length_5513_cov_46.199_g556_i0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_807_length_5513_cov_37.6629_g556_i1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       name  tax_id  rank  tax_name top_label\n",
       "0   NODE_646_length_5889_cov_28.079_g448_i0       1     1         1       N/A\n",
       "1  NODE_667_length_5857_cov_28.2491_g448_i1       1     1         1       N/A\n",
       "2  NODE_671_length_5843_cov_18.2623_g465_i0       1     1         1       N/A\n",
       "3   NODE_806_length_5513_cov_46.199_g556_i0       1     1         1       N/A\n",
       "4  NODE_807_length_5513_cov_37.6629_g556_i1       1     1         1       N/A"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "megan_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table = pd.merge(result,megan_annot, on=['name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>probability</th>\n",
       "      <th>prediction</th>\n",
       "      <th>vfam</th>\n",
       "      <th>pfam</th>\n",
       "      <th>diamond</th>\n",
       "      <th>label</th>\n",
       "      <th>score_euk_non_euk</th>\n",
       "      <th>pred_euk_non_euk</th>\n",
       "      <th>score_non_euk_virus</th>\n",
       "      <th>pred_non_euk_virus</th>\n",
       "      <th>score_virus_non_virus</th>\n",
       "      <th>pred_virus_non_virus</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tax_name</th>\n",
       "      <th>top_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_1_length_91451_cov_163.114_g0_i0</td>\n",
       "      <td>91451</td>\n",
       "      <td>0.000375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996004</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001954</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1236</td>\n",
       "      <td>1236</td>\n",
       "      <td>1236</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_2_length_74096_cov_303.895_g1_i0</td>\n",
       "      <td>74096</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.995278</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000529</td>\n",
       "      <td>0.0</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_3_length_66420_cov_268.59_g1_i1</td>\n",
       "      <td>66420</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996629</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000761</td>\n",
       "      <td>0.0</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_4_length_61338_cov_317.585_g2_i0</td>\n",
       "      <td>61338</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.991967</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_5_length_59036_cov_197.701_g0_i1</td>\n",
       "      <td>59036</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.990354</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.0</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    name  length  probability  prediction  \\\n",
       "0  NODE_1_length_91451_cov_163.114_g0_i0   91451     0.000375         0.0   \n",
       "1  NODE_2_length_74096_cov_303.895_g1_i0   74096     0.000088         0.0   \n",
       "2   NODE_3_length_66420_cov_268.59_g1_i1   66420     0.000041         0.0   \n",
       "3  NODE_4_length_61338_cov_317.585_g2_i0   61338     0.000056         0.0   \n",
       "4  NODE_5_length_59036_cov_197.701_g0_i1   59036     0.000040         0.0   \n",
       "\n",
       "   vfam  pfam  diamond  label  score_euk_non_euk  pred_euk_non_euk  \\\n",
       "0   0.0   0.0      0.0    0.0           0.000025               0.0   \n",
       "1   0.0   0.0      0.0    0.0           0.000012               0.0   \n",
       "2   0.0   0.0      0.0    0.0           0.000009               0.0   \n",
       "3   0.0   0.0      0.0    0.0           0.000008               0.0   \n",
       "4   0.0   0.0      0.0    0.0           0.000008               0.0   \n",
       "\n",
       "   score_non_euk_virus  pred_non_euk_virus  score_virus_non_virus  \\\n",
       "0             0.996004                 1.0               0.001954   \n",
       "1             0.995278                 1.0               0.000529   \n",
       "2             0.996629                 1.0               0.000761   \n",
       "3             0.991967                 1.0               0.000477   \n",
       "4             0.990354                 1.0               0.000617   \n",
       "\n",
       "   pred_virus_non_virus  tax_id    rank  tax_name top_label  \n",
       "0                   0.0    1236    1236      1236  Bacteria  \n",
       "1                   0.0  254161  254161    254161  Bacteria  \n",
       "2                   0.0  254161  254161    254161  Bacteria  \n",
       "3                   0.0     286     286       286  Bacteria  \n",
       "4                   0.0     286     286       286  Bacteria  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_300k = pd.read_csv('/global/projectb/scratch/qpzhang/TARA/Libsvm/ViCA2/all_2k_contigs.fa.prediction.Adam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_adam = pd.read_csv('/global/projectb/scratch/qpzhang/TARA/Libsvm/ViCA2/all_2k_contigs.fa.prediction.1M_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Class</th>\n",
       "      <th>Class_id</th>\n",
       "      <th>Prob_class_0</th>\n",
       "      <th>Prob_class_1</th>\n",
       "      <th>Prob_class_2</th>\n",
       "      <th>Prob_class_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_10000_length_2000_cov_95.909_g6946_i0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.021495</td>\n",
       "      <td>1.915236e-03</td>\n",
       "      <td>9.331824e-04</td>\n",
       "      <td>0.975657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_10000_length_2048_cov_10.8329_g7692_i0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.005805</td>\n",
       "      <td>5.705581e-05</td>\n",
       "      <td>1.474175e-04</td>\n",
       "      <td>0.993991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_10000_length_2108_cov_40.585_g7688_i0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.378866e-09</td>\n",
       "      <td>1.581349e-08</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_10000_length_2114_cov_17.7008_g7479_i0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>8.563400e-05</td>\n",
       "      <td>1.848443e-05</td>\n",
       "      <td>0.999644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_10000_length_2132_cov_22.4521_g7241_i0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>9.539825e-01</td>\n",
       "      <td>7.859741e-08</td>\n",
       "      <td>0.046000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            ID  Class  Class_id  Prob_class_0  \\\n",
       "0   NODE_10000_length_2000_cov_95.909_g6946_i0      3         3      0.021495   \n",
       "1  NODE_10000_length_2048_cov_10.8329_g7692_i0      3         3      0.005805   \n",
       "2   NODE_10000_length_2108_cov_40.585_g7688_i0      3         3      0.000002   \n",
       "3  NODE_10000_length_2114_cov_17.7008_g7479_i0      3         3      0.000252   \n",
       "4  NODE_10000_length_2132_cov_22.4521_g7241_i0      1         1      0.000017   \n",
       "\n",
       "   Prob_class_1  Prob_class_2  Prob_class_3  \n",
       "0  1.915236e-03  9.331824e-04      0.975657  \n",
       "1  5.705581e-05  1.474175e-04      0.993991  \n",
       "2  1.378866e-09  1.581349e-08      0.999998  \n",
       "3  8.563400e-05  1.848443e-05      0.999644  \n",
       "4  9.539825e-01  7.859741e-08      0.046000  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prediction_adam.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">Class_id</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Prob_class_0</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Prob_class_2</th>\n",
       "      <th colspan=\"8\" halign=\"left\">Prob_class_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>...</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Class</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25984.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25984.0</td>\n",
       "      <td>0.724670</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022200</td>\n",
       "      <td>0.157081</td>\n",
       "      <td>25984.0</td>\n",
       "      <td>0.175443</td>\n",
       "      <td>0.133476</td>\n",
       "      <td>6.935089e-05</td>\n",
       "      <td>5.727035e-02</td>\n",
       "      <td>0.147544</td>\n",
       "      <td>0.278886</td>\n",
       "      <td>0.497957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27221.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>27221.0</td>\n",
       "      <td>0.023249</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.119974</td>\n",
       "      <td>27221.0</td>\n",
       "      <td>0.070493</td>\n",
       "      <td>0.119092</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.914311e-04</td>\n",
       "      <td>0.007034</td>\n",
       "      <td>0.082659</td>\n",
       "      <td>0.499408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.019477</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.045909</td>\n",
       "      <td>0.088688</td>\n",
       "      <td>6.078228e-17</td>\n",
       "      <td>2.764050e-07</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.019434</td>\n",
       "      <td>0.278858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>160358.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>160358.0</td>\n",
       "      <td>0.038752</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001210</td>\n",
       "      <td>0.243533</td>\n",
       "      <td>160358.0</td>\n",
       "      <td>0.937465</td>\n",
       "      <td>0.126815</td>\n",
       "      <td>3.230318e-01</td>\n",
       "      <td>9.555361e-01</td>\n",
       "      <td>0.997928</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Class_id                                    Prob_class_0            \\\n",
       "          count mean  std  min  25%  50%  75%  max        count      mean   \n",
       "Class                                                                       \n",
       "0       25984.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0      25984.0  0.724670   \n",
       "1       27221.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0      27221.0  0.023249   \n",
       "2          21.0  2.0  0.0  2.0  2.0  2.0  2.0  2.0         21.0  0.019477   \n",
       "3      160358.0  3.0  0.0  3.0  3.0  3.0  3.0  3.0     160358.0  0.038752   \n",
       "\n",
       "         ...    Prob_class_2           Prob_class_3                      \\\n",
       "         ...             75%       max        count      mean       std   \n",
       "Class    ...                                                              \n",
       "0        ...        0.022200  0.157081      25984.0  0.175443  0.133476   \n",
       "1        ...        0.000012  0.119974      27221.0  0.070493  0.119092   \n",
       "2        ...        1.000000  1.000000         21.0  0.045909  0.088688   \n",
       "3        ...        0.001210  0.243533     160358.0  0.937465  0.126815   \n",
       "\n",
       "                                                                 \n",
       "                min           25%       50%       75%       max  \n",
       "Class                                                            \n",
       "0      6.935089e-05  5.727035e-02  0.147544  0.278886  0.497957  \n",
       "1      0.000000e+00  3.914311e-04  0.007034  0.082659  0.499408  \n",
       "2      6.078228e-17  2.764050e-07  0.000036  0.019434  0.278858  \n",
       "3      3.230318e-01  9.555361e-01  0.997928  0.999885  1.000000  \n",
       "\n",
       "[4 rows x 40 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prediction_adam.groupby('Class').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_adam_subset = prediction_adam[['ID','Prob_class_0']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/homes/q/qpzhang/vica/miniconda/envs/vica/lib/python3.6/site-packages/pandas/core/frame.py:3781: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  return super(DataFrame, self).rename(**kwargs)\n"
     ]
    }
   ],
   "source": [
    "prediction_adam_subset.rename(index=str,columns={\"ID\": \"name\"},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_adam_subset.rename(index=str,columns={\"Prob_class_0\": \"1M_prediction\"},inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_300k_subset = prediction_300k[['ID','Prob_class_0']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_300k_subset.rename(index=str,columns={\"ID\": \"name\"},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_300k_subset.rename(index=str,columns={\"Prob_class_0\": \"100k_prediction\"},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table_300k_joined = pd.merge(new_table,prediction_300k_subset, how='left',on=['name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table_300k_adam_joined = pd.merge(new_table_300k_joined,prediction_adam_subset, how='left',on=['name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7041 contigs as ground-truth virus-related, with hits to virus-related domains, or annotated as virus by MEGAN based on BlastX/Diamond alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 208302 contigs with length >=2k, in TARA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table_300k_adam_joined[\"non_euk_label\"] = (new_table_300k_adam_joined.top_label!='Eukaryota')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>probability</th>\n",
       "      <th>prediction</th>\n",
       "      <th>vfam</th>\n",
       "      <th>pfam</th>\n",
       "      <th>diamond</th>\n",
       "      <th>label</th>\n",
       "      <th>score_euk_non_euk</th>\n",
       "      <th>pred_euk_non_euk</th>\n",
       "      <th>...</th>\n",
       "      <th>pred_non_euk_virus</th>\n",
       "      <th>score_virus_non_virus</th>\n",
       "      <th>pred_virus_non_virus</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tax_name</th>\n",
       "      <th>top_label</th>\n",
       "      <th>100k_prediction</th>\n",
       "      <th>1M_prediction</th>\n",
       "      <th>non_euk_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_1_length_91451_cov_163.114_g0_i0</td>\n",
       "      <td>91451</td>\n",
       "      <td>0.000375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001954</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1236</td>\n",
       "      <td>1236</td>\n",
       "      <td>1236</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>9.194513e-19</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_2_length_74096_cov_303.895_g1_i0</td>\n",
       "      <td>74096</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000529</td>\n",
       "      <td>0.0</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>4.935380e-19</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_3_length_66420_cov_268.59_g1_i1</td>\n",
       "      <td>66420</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000761</td>\n",
       "      <td>0.0</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>254161</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>4.259754e-14</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_4_length_61338_cov_317.585_g2_i0</td>\n",
       "      <td>61338</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>9.107938e-19</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_5_length_59036_cov_197.701_g0_i1</td>\n",
       "      <td>59036</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.0</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>6.936213e-15</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    name  length  probability  prediction  \\\n",
       "0  NODE_1_length_91451_cov_163.114_g0_i0   91451     0.000375         0.0   \n",
       "1  NODE_2_length_74096_cov_303.895_g1_i0   74096     0.000088         0.0   \n",
       "2   NODE_3_length_66420_cov_268.59_g1_i1   66420     0.000041         0.0   \n",
       "3  NODE_4_length_61338_cov_317.585_g2_i0   61338     0.000056         0.0   \n",
       "4  NODE_5_length_59036_cov_197.701_g0_i1   59036     0.000040         0.0   \n",
       "\n",
       "   vfam  pfam  diamond  label  score_euk_non_euk  pred_euk_non_euk  \\\n",
       "0   0.0   0.0      0.0    0.0           0.000025               0.0   \n",
       "1   0.0   0.0      0.0    0.0           0.000012               0.0   \n",
       "2   0.0   0.0      0.0    0.0           0.000009               0.0   \n",
       "3   0.0   0.0      0.0    0.0           0.000008               0.0   \n",
       "4   0.0   0.0      0.0    0.0           0.000008               0.0   \n",
       "\n",
       "       ...        pred_non_euk_virus  score_virus_non_virus  \\\n",
       "0      ...                       1.0               0.001954   \n",
       "1      ...                       1.0               0.000529   \n",
       "2      ...                       1.0               0.000761   \n",
       "3      ...                       1.0               0.000477   \n",
       "4      ...                       1.0               0.000617   \n",
       "\n",
       "   pred_virus_non_virus  tax_id    rank  tax_name  top_label 100k_prediction  \\\n",
       "0                   0.0    1236    1236      1236   Bacteria        0.000245   \n",
       "1                   0.0  254161  254161    254161   Bacteria        0.000019   \n",
       "2                   0.0  254161  254161    254161   Bacteria        0.000052   \n",
       "3                   0.0     286     286       286   Bacteria        0.000032   \n",
       "4                   0.0     286     286       286   Bacteria        0.000037   \n",
       "\n",
       "   1M_prediction  non_euk_label  \n",
       "0   9.194513e-19           True  \n",
       "1   4.935380e-19           True  \n",
       "2   4.259754e-14           True  \n",
       "3   9.107938e-19           True  \n",
       "4   6.936213e-15           True  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_table_300k_adam_joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/homes/q/qpzhang/vica/miniconda/envs/vica/lib/python3.6/site-packages/pandas/core/ops.py:1167: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  result = method(y)\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "invalid type comparison",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-45f6c505ca09>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m new_table_300k_adam_joined[\"micromonas_label\"] = (new_table_300k_adam_joined.tax_name=='Micromonas') |(new_table_300k_adam_joined.tax_name=='Micromonas pusilla CCMP1545')|(\n\u001b[0m\u001b[1;32m      2\u001b[0m         new_table_300k_adam_joined.tax_name=='Micromonas commoda')\n",
      "\u001b[0;32m~/vica/miniconda/envs/vica/lib/python3.6/site-packages/pandas/core/ops.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(self, other, axis)\u001b[0m\n\u001b[1;32m   1281\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1282\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merrstate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'ignore'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1283\u001b[0;31m                 \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mna_op\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1284\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1285\u001b[0m                 raise TypeError('Could not compare {typ} type with Series'\n",
      "\u001b[0;32m~/vica/miniconda/envs/vica/lib/python3.6/site-packages/pandas/core/ops.py\u001b[0m in \u001b[0;36mna_op\u001b[0;34m(x, y)\u001b[0m\n\u001b[1;32m   1167\u001b[0m                     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1168\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNotImplemented\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1169\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"invalid type comparison\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1170\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1171\u001b[0m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: invalid type comparison"
     ]
    }
   ],
   "source": [
    "new_table_300k_adam_joined[\"micromonas_label\"] = (new_table_300k_adam_joined.tax_name=='Micromonas') |(new_table_300k_adam_joined.tax_name=='Micromonas pusilla CCMP1545')|(\n",
    "        new_table_300k_adam_joined.tax_name=='Micromonas commoda')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### \"no rank\" does not equal \"no annotation\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_top_label(top_label):\n",
    "    if top_label == 'N/A':\n",
    "        return 'no annotation'\n",
    "    else:\n",
    "        return 'with annotation'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table_300k_adam_joined['annotation_label'] = new_table_300k_adam_joined['top_label'].map(convert_top_label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split the TARA contigs into 3 categories\n",
    "\n",
    "- \"Virus\" has virus-related domain hits or annotated as \"virus\" by MEGAN\n",
    "- \"Non-virus\" , no virus-related domain hits, annotated as some non-virus taxonomical group by MEGAN \n",
    "- \"Unknown\",  no virus-related domain hits, not annotated into any taxonomical group by MEGAN, top_label as \"N/A\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### subsample \"non-virus\" and \"unknown\" to have same number of \"Virus\" contigs - 6913"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"non-virus\" data frame... non-virus and with annotation by MEGAN\n",
    "df_nonvirus = new_table_300k_adam_joined[new_table_300k_adam_joined.label==0.0][new_table_300k_adam_joined['annotation_label']== 'with annotation']\n",
    "df_nonvirus_subsample = df_nonvirus.sample(n=7041, replace=False)\n",
    "#df_nonvirus_subsample.to_csv('df_nonvirus_subsample.csv')\n",
    "\n",
    "# \"Unknown\" data frame... and without annotation by MEGAN\n",
    "\n",
    "df_unknown = new_table_300k_adam_joined[new_table_300k_adam_joined.label==0.0][new_table_300k_adam_joined['annotation_label']== 'no annotation']\n",
    "df_unknown_subsample = df_unknown.sample(n=7041, replace=False)\n",
    "#df_unknown_subsample.to_csv('df_unknown_subsample.csv')\n",
    "\n",
    "### \"Virus\" contigs...  has pfam hits or annotated as \"virus\" by MEGAN\n",
    "df_virus = new_table_300k_adam_joined[new_table_300k_adam_joined.label==1.0]\n",
    "#df_virus.to_csv('df_virus.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_category(df):\n",
    "    if df.label == 0.0:\n",
    "        if df.annotation_label == 'with annotation':\n",
    "            return 'nonvirus'\n",
    "        else:\n",
    "            return 'unknown'\n",
    "    else:\n",
    "        return 'virus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table_300k_adam_joined['category'] = new_table_300k_adam_joined.apply(get_category, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_table_300k_adam_joined.groupby('category').describe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_micromonus = new_table_300k_adam_joined[new_table_300k_adam_joined.micromonas_label==True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Micromonus related contigs\n",
    "\n",
    "### updated model with 100K segments by Adam\n",
    "\n",
    "New model has better performance to label Micromonus as \"non-virus\" - lower score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_micromonus[\"1M_prediction\"].plot.hist(bins=50, title='adam model, for micromonus related contigs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Older model with 300K segments by QP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_micromonus['100k_prediction'].plot.hist(bins=50, title='300k old model, for micromonus related contigs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All TARA contigs, with updated model  using 100K training segments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(16,9))\n",
    "ax0 = fig.add_subplot(111)\n",
    "n_bins = 30\n",
    "#colors = ['red', 'blue', 'grey']\n",
    "x_multi = [new_table_300k_adam_joined[new_table_300k_adam_joined['1M_prediction']>0]['1M_prediction']]\n",
    "ax0.hist(x_multi, n_bins, histtype='bar')\n",
    "ax0.set_title('All TARA contigas, Tensorflow updated model with 100K training segments',fontsize=24)\n",
    "ax0.set_xlabel('score',fontsize=24)\n",
    "ax0.tick_params(labelsize=20)\n",
    "ax0.set_ylabel('number of contigs',fontsize=24)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## \"nonvirus\" TARA contigs, with updated model using 100K training segments "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(16,9))\n",
    "ax0 = fig.add_subplot(111)\n",
    "n_bins = 100\n",
    "#colors = ['red', 'blue', 'grey']\n",
    "x_multi = [df_nonvirus[df_nonvirus['1M_prediction']>0]['1M_prediction']]\n",
    "ax0.hist(x_multi, n_bins, histtype='bar',  label=[\n",
    "                                                  'non-virus'],\n",
    "         color=['grey'])\n",
    "ax0.set_title('\"nonvirus\" TARA contigs >2K bps, Tensorflow updated model with 100K training segments',fontsize=24)\n",
    "ax0.set_xlabel('score',fontsize=24)\n",
    "ax0.tick_params(labelsize=20)\n",
    "ax0.set_ylabel('number of contigs',fontsize=24)\n",
    "ax0.legend(prop={'size': 24})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## \"virus\" TARA contigs, with updated model using 100K training segments "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(16,9))\n",
    "ax0 = fig.add_subplot(111)\n",
    "n_bins = 20\n",
    "#colors = ['red', 'blue', 'grey']\n",
    "x_multi = [df_virus[df_virus[\"1M_prediction\"]>0][\"1M_prediction\"]]\n",
    "ax0.hist(x_multi, n_bins, histtype='bar',  label=[\n",
    "                                                  'virus'],\n",
    "         color=['goldenrod'])\n",
    "ax0.set_title('\"virus\" TARA contigs >2K bps, Tensorflow updated model with 100K training segments',fontsize=24)\n",
    "ax0.set_xlabel('score',fontsize=24)\n",
    "ax0.tick_params(labelsize=20)\n",
    "ax0.set_ylabel('number of contigs',fontsize=24)\n",
    "ax0.legend(prop={'size': 24})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check the distribution of probability of same number of \"virus\", \"non-virus\", 'unknown\" contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Balanced data, same number of \"virus\" and \"non-virus\" contigs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(16,9))\n",
    "ax0 = fig.add_subplot(111)\n",
    "n_bins = 10\n",
    "#colors = ['red', 'blue', 'grey']\n",
    "x_multi = [df_nonvirus_subsample.adam_prediction,df_virus.adam_prediction, df_unknown_subsample.adam_prediction ]\n",
    "ax0.hist(x_multi, n_bins, histtype='bar',  label=['non-virus','virus','unknown'],color=['grey', 'goldenrod', 'green'])\n",
    "ax0.set_title('On balanced subsets of TARA contigs, Tensorflow updated model with 100K training segments',fontsize=24)\n",
    "ax0.set_xlabel('score',fontsize=24)\n",
    "ax0.tick_params(labelsize=20)\n",
    "ax0.set_ylabel('number of contigs',fontsize=24)\n",
    "ax0.legend(prop={'size': 24})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Balanced TARA contigs (same number of \"virus\" and \"non-virus\")\n",
    "\n",
    "### comparing the updated model with 100K training segments and old model with 300K training segments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(16,9))\n",
    "ax0 = fig.add_subplot(111)\n",
    "n_bins = 10\n",
    "#colors = ['red', 'blue', 'grey']\n",
    "x_multi = [df_nonvirus_subsample['300k_prediction'],df_nonvirus_subsample.adam_prediction,df_virus['300k_prediction'],\n",
    "           df_virus.adam_prediction,df_unknown_subsample['300k_prediction'],df_unknown_subsample.adam_prediction ]\n",
    "ax0.hist(x_multi, n_bins, histtype='bar',  label=['non-virus_old_300k','non-virus-updated_100k',\n",
    "                                                  'virus_old_300k','virus_updated_100k',\n",
    "                                                  'unknown_old_300k','unknown_updated_100k'],\n",
    "         color=['lightgrey','grey', 'wheat','goldenrod', 'lightgreen','green'])\n",
    "ax0.set_title('On balanced subsets of TARA contigs',fontsize=24)\n",
    "ax0.set_xlabel('score',fontsize=24)\n",
    "ax0.set_ylabel('number of contigs',fontsize=24)\n",
    "ax0.tick_params(labelsize=20)\n",
    "ax0.legend(prop={'size': 24})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "VICA",
   "language": "python",
   "name": "vica"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
