<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to GeneLearn’s documentation! &mdash; GeneLearn 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="GeneLearn 0.1 documentation" href="#" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="#">GeneLearn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-genelearn-s-documentation">
<h1>Welcome to GeneLearn&#8217;s documentation!<a class="headerlink" href="#welcome-to-genelearn-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
<p>A program to classify, bin and discover divergent  genomic sequences with machine learning using novel feature extraction methods</p>
<div class="section" id="features">
<h2>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Classifying metagenomic and metatransciptomic contigs taxonomically</li>
<li>Creating data files for genome binning using emergent self organizing maps (ESOM)</li>
<li>Identifying metagenomic contigs and metatransciptomic contigs that come from deeply
divergent forms of life</li>
</ul>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>Dependencies:</p>
<ul class="simple">
<li>GenemarkS version 4.29  (<a class="reference external" href="http://exon.gatech.edu/GeneMark/">http://exon.gatech.edu/GeneMark/</a>)</li>
<li>RefTree  (<a class="reference external" href="https://bitbucket.org/berkeleylab/jgi_reftree">https://bitbucket.org/berkeleylab/jgi_reftree</a>)</li>
<li>Task Farmer  (<a class="reference external" href="http://jgi.goe.gov">http://jgi.goe.gov</a>)</li>
<li>Python v2.74  (<a class="reference external" href="https://www.python.org/">https://www.python.org/</a>)</li>
<li>scikit-learn  (<a class="reference external" href="http://scikit-learn.org/">http://scikit-learn.org/</a>)</li>
<li>Biopython  (<a class="reference external" href="http://biopython.org">http://biopython.org</a>)</li>
<li>simplejson  (<a class="reference external" href="https://github.com/simplejson/simplejson">https://github.com/simplejson/simplejson</a>)</li>
<li>NumPy  (<a class="reference external" href="http://www.numpy.org/">http://www.numpy.org/</a>)</li>
<li>SciPy  (<a class="reference external" href="http://www.scipy.org/">http://www.scipy.org/</a>)</li>
<li>Matplotlib  (<a class="reference external" href="http://matplotlib.org/">http://matplotlib.org/</a>)</li>
</ul>
</div>
<div class="section" id="files">
<h2>Files<a class="headerlink" href="#files" title="Permalink to this headline">¶</a></h2>
<div class="section" id="setup">
<h3>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul class="simple">
<li>env.sh - source environment variables</li>
<li>gm_parameters A directory with modified GeneMark Parameter files</li>
<li>config.json, training.json - config files for running the classifier</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="creating-training-data">
<h3>Creating Training data<a class="headerlink" href="#creating-training-data" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">crate_training_data.py - a wrapper to have reftree run single_taxon_training.py on each taxon below the selected taxonomic node</p>
</li>
<li><p class="first">single_taxon_training.sh - a shell script that calls single_taxon_training.py</p>
</li>
<li><p class="first">single_taxon_training.py - A wrapper to have run tasks necessary to create training data for a single genome</p>
<blockquote>
<div><ol class="arabic simple">
<li>shred.py - shreds the genome into smaller pieces of fixed length or sizes conforming to a gamma distribution</li>
<li>Feature extraction script - see below of a list of features</li>
</ol>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="model-training-and-cross-validation-via-svm">
<h3>model training and cross validation via SVM<a class="headerlink" href="#model-training-and-cross-validation-via-svm" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">svm_training_cross_validation.py</p>
<blockquote>
<div><ol class="arabic simple">
<li>training_data_formatter.py - a script to take taxonomic nodes or levels and create a multi feature JSON file for classification</li>
<li>svm_training.py - A script to take training data, do a parameter search, create a model and cross-validate it</li>
</ol>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="sample-classification">
<h3>Sample classification<a class="headerlink" href="#sample-classification" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">svm_classification.py - a wrapper to run SVM classification</p>
<blockquote>
<div><ol class="arabic simple">
<li>feature extraction scripts - see below</li>
<li>feature_formatter.py - a utility script  to format csv feature data into json data</li>
<li>svm_classifier.py a script to load a SVM model and classify sequences</li>
</ol>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="producing-data-for-genome-binning">
<h3>Producing data for genome binning<a class="headerlink" href="#producing-data-for-genome-binning" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">metagenome_binning_data.py -a a wrapper for producing genome binning data</p>
<blockquote>
<div><ol class="arabic simple">
<li>Feature extraction script</li>
<li>feature formatter</li>
</ol>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="feature-extraction-ecripts">
<h3>Feature extraction ecripts<a class="headerlink" href="#feature-extraction-ecripts" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>feature_extraction_metamark.py - a wrapper to run metamark and extract the genomic feature</li>
<li>feature_extraction_kmer.py an alternate method of extracting feature information by kmer</li>
</ul>
</div>
<div class="section" id="utility-scripts">
<h3>Utility scripts<a class="headerlink" href="#utility-scripts" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>feature_formatter.py - a script to convert between tab delimited and ESOM data formats</li>
<li>estimate_gamma_parameters.py - a script to estimate gamma parameters from a dataset that will be classified</li>
</ul>
</div>
</div>
<div class="section" id="diagram-of-program">
<h2>Diagram of program<a class="headerlink" href="#diagram-of-program" title="Permalink to this headline">¶</a></h2>
<img alt="_images/GeneLearn.png" src="_images/GeneLearn.png" />
</div>
<div class="section" id="support">
<h2>Support<a class="headerlink" href="#support" title="Permalink to this headline">¶</a></h2>
<p>If you are having issues, please let us know.
We have a mailing list located at: <a class="reference external" href="mailto:project&#37;&#52;&#48;google-groups&#46;com">project<span>&#64;</span>google-groups<span>&#46;</span>com</a></p>
</div>
<div class="section" id="license">
<h2>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h2>
<p>The project is licensed under the XXX license.</p>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span>Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to GeneLearn&#8217;s documentation!</a><ul>
<li><a class="reference internal" href="#features">Features</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#files">Files</a><ul>
<li><a class="reference internal" href="#setup">Setup</a></li>
<li><a class="reference internal" href="#creating-training-data">Creating Training data</a></li>
<li><a class="reference internal" href="#model-training-and-cross-validation-via-svm">model training and cross validation via SVM</a></li>
<li><a class="reference internal" href="#sample-classification">Sample classification</a></li>
<li><a class="reference internal" href="#producing-data-for-genome-binning">Producing data for genome binning</a></li>
<li><a class="reference internal" href="#feature-extraction-ecripts">Feature extraction ecripts</a></li>
<li><a class="reference internal" href="#utility-scripts">Utility scripts</a></li>
</ul>
</li>
<li><a class="reference internal" href="#diagram-of-program">Diagram of program</a></li>
<li><a class="reference internal" href="#support">Support</a></li>
<li><a class="reference internal" href="#license">License</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Adam Rivers, Qingpeng Zhang, Edward Eirton.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.3</a>
      
      |
      <a href="_sources/index.txt"
          rel="nofollow">Page source</a></li>
    </div>

    

    
  </body>
</html>